### ISSUE TITLE
DistributedSampler can't shuffle the dataset

### ISSUE THREAD

**Issue by kangcp** [pytorch/pytorch#31771]
## 🐛 Bug

When using the torch.utils.data.distributed.DistributedSampler () function, the data set cannot be shuffled, and each epoch samples the data in the same order.

## To Reproduce

```python
import torch
import os
from torch.utils.data import Dataset, DataLoader
import numpy as np

os.environ["CUDA_VISIBLE_DEVICES"] = "0"

np.random.seed(5)

class RandomDataset(Dataset):
    def __init__(self):
        self.data = torch.rand(22, 1, 32, 32)
        self.name = torch.arange(1, 22)
        
    def __getitem__(self, idx):
        return self.name[idx], self.data[idx]
    
    def __len__(self):
        return len(self.data)
    

torch.distributed.init_process_group(backend="nccl")

dataset = RandomDataset()
sampler = torch.utils.data.distributed.DistributedSampler(dataset, shuffle=True) 
dataloader = DataLoader(dataset, 
    batch_size=5,   
    pin_memory=True, drop_last=True, sampler=sampler) 

for epoch in range(3):
    print("epoch: ", epoch)
    for i, data in enumerate(dataloader, 0):
        names, _ = data
        print(names)
```

Run this code by executing:
```
python -m torch.distributed.launch --nproc_per_node=1 test.py
```

## Expected behavior
Expect different data batch combinations for each epoch.
for example:
In epoch 1, My code prints:
tensor([10,  4, 15,  3, 11])
tensor([ 5,  9, 20, 19, 13])
tensor([ 2,  0, 18, 16,  7])
tensor([ 8,  6, 14, 17,  1])
In epoch 2, My code prints:
tensor([ 5,  9, 15,  3, 13])
tensor([10,  4, 20, 19, 11])
tensor([ 2,  14, 17, 0,  1])
tensor([ 8,  6, 18, 16,  7])
and so on.

But now, the output is:
epoch:  1
tensor([10,  4, 15,  3, 11])
tensor([ 5,  9, 20, 19, 13])
tensor([ 2,  0, 18, 16,  7])
tensor([ 8,  6, 14, 17,  1])
epoch:  2
tensor([10,  4, 15,  3, 11])
tensor([ 5,  9, 20, 19, 13])
tensor([ 2,  0, 18, 16,  7])
tensor([ 8,  6, 14, 17,  1])

Every epoch, the names are the same。

## Environment

PyTorch version: 1.3.1+cu100
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 6.5.0-2ubuntu1~16.04) 6.5.0 20181026
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration:
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti
GPU 2: GeForce GTX 1080 Ti
GPU 3: GeForce GTX 1080 Ti

Nvidia driver version: 418.67
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.15.4
[conda] _tflow_select             2.3.0                       mkl
[conda] pytorch-wpe               0.0.0                    pypi_0    pypi
[conda] tensorflow                1.14.0          mkl_py36h2526735_0
[conda] tensorflow-base           1.14.0          mkl_py36h7ce6ba3_0
[conda] torch                     1.3.1+cu100              pypi_0    pypi
[conda] torch-complex             0.0.2                    pypi_0    pypi
[conda] warpctc-pytorch           0.1                      pypi_0    pypi
[conda] warpctc-pytorch10-cuda100 0.1.3                    pypi_0    pypi
[conda] warprnnt-pytorch          0.1                      pypi_0    pypi

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

## Additional context

cc @ezyang @gchanan @zou3519 @SsnL @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @xush6528

**Comment by ezyang** [comment]
I am not too sure about the severity of this issue, but it seems that it could be quite severe (silent correctness problem). I have raised the priority pending a look from someone more familiar.

**Comment by kangcp** [comment]
> I am not too sure about the severity of this issue, but it seems that it could be quite severe (silent correctness problem). I have raised the priority pending a look from someone more familiar.

ok，Thank you.

**Comment by kangcp** [comment]
@ezyang Hi,  I think I have found the bug.
https://pytorch.org/docs/stable/_modules/torch/utils/data/distributed.html#DistributedSampler.
This is the source code of DistributedSampler. 
In the __iter__()  function, there are two lines of code: 
```python
g = torch.Generator ()
g.manual_seed (self.epoch)
```
So in each epoch, A new Generator object is generated every time, and the seed is the same every time. Therefore, each epoch will sample the data set in the same order. The solution to solve this problem is to turn the Generator into a global property, such as put 
```python
g = torch.Generator ()
g.manual_seed (self.epoch) 
```
in the __init__() function.

**Comment by kangcp** [comment]
@ezyang  I found a set_epoch() function in the DistributedSampler source code, but I haven't found any place to call this function. This needs to be called manually by the user? I didn't find a description in the doc file, and there was no description of this function in the doc file.

**Comment by vfdev-5** [comment]
@kangcp you need to call `set_epoch` as here: https://github.com/pytorch/examples/blob/master/imagenet/main.py#L239

**Comment by kangcp** [comment]
> @kangcp you need to call `set_epoch` as here: https://github.com/pytorch/examples/blob/master/imagenet/main.py#L239

Thank you, I have call that.

**Comment by pritamdamania87** [comment]
@kangcp Can this issue be closed out?

**Comment by rgommers** [comment]
This requires at least a documentation update before the issue can be closed.

There's also an implementation issue, `g.manual_seed(self.epoch)` inside `DistributedSampler` is a very low-entropy way to seed. The `manual_seed` docstring recommends against this: *It is recommended to set a large seed, i.e. a number that has a good balance of 0 and 1 bits. Avoid having many 0 bits in the seed.* It may not be possible to change it anymore due to backwards compatibility, but I'll have a look if there's something that can be done or to warn the user.
